backend/utils/bulk_import.py

```
@@ -189,15 +198,24 @@ def validate_chemical_data(row_data: Dict[str, Any]) -> Dict[str, Any]:
+            chemical_data['warehouse_id'] = int(warehouse_id_str)
+        except ValueError:
+            raise ValidationError(f"Invalid warehouse_id: {warehouse_id_str}")
+
+    # Handle quantity - must be integer
     quantity_str = row_data.get('quantity', '0')
     try:
-        chemical_data['quantity'] = float(quantity_str) if quantity_str else 0
+        chemical_data['quantity'] = int(float(quantity_str)) if quantity_str else 0
     except ValueError:
-        raise ValidationError(f"Invalid quantity: {quantity_str}")
+        raise ValidationError(f"Invalid quantity: {quantity_str}. Quantity must be a whole number.")
```

Ruff (0.14.1)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)